File: native/iOS/ProduceApp/Data/ProduceService.swift
import Foundation

actor ProduceService {
    static let shared = ProduceService()
    private let baseUrl = Configuration.apiBaseUrl

    // MARK: - JWT Management

    func ensureJwtToken() async {
        if UserDefaults.standard.string(forKey: "jwt_token") != nil { return }

        let deviceId = UIDevice.current.identifierForVendor?.uuidString ?? UUID().uuidString
        guard let url = URL(string: "\(baseUrl)auth/token") else { return }

        var request = URLRequest(url: url)
        request.httpMethod = "POST"
        request.setValue("application/json", forHTTPHeaderField: "Content-Type")
        request.httpBody = try? JSONEncoder().encode(["deviceId": deviceId])

        do {
            let (data, _) = try await URLSession.shared.data(for: request)
            if let json = try? JSONSerialization.jsonObject(with: data) as? [String: Any],
               let token = json["token"] as? String {
                UserDefaults.standard.set(token, forKey: "jwt_token")
            }
        } catch {
            print("JWT 取得失敗: \(error)")
        }
    }

    private func authorizedRequest(url: URL) -> URLRequest {
        var request = URLRequest(url: url)
        if let token = UserDefaults.standard.string(forKey: "jwt_token") {
            request.setValue("Bearer \(token)", forHTTPHeaderField: "Authorization")
        }
        return request
    }

    // MARK: - API Methods

    func fetchDailyPrices(keyword: String? = nil, page: Int = 1) async throws -> [ProduceDto] {
        var components = URLComponents(string: "\(baseUrl)daily-prices")!
        var queryItems = [URLQueryItem(name: "page", value: "\(page)")]
        if let keyword = keyword, !keyword.isEmpty {
            queryItems.append(URLQueryItem(name: "keyword", value: keyword))
        }
        components.queryItems = queryItems

        let request = authorizedRequest(url: components.url!)
        let (data, _) = try await URLSession.shared.data(for: request)
        return try JSONDecoder().decode([ProduceDto].self, from: data)
    }

    func fetchPriceHistory(produceId: String) async throws -> [HistoricalPriceDto] {
        let url = URL(string: "\(baseUrl)history/\(produceId)")!
        let request = authorizedRequest(url: url)
        let (data, _) = try await URLSession.shared.data(for: request)
        return try JSONDecoder().decode([HistoricalPriceDto].self, from: data)
    }

    func fetchTopVolume() async throws -> [ProduceDto] {
        let url = URL(string: "\(baseUrl)top-volume")!
        let request = authorizedRequest(url: url)
        let (data, _) = try await URLSession.shared.data(for: request)
        return try JSONDecoder().decode([ProduceDto].self, from: data)
    }

    func fetchAnomalies() async throws -> [PriceAnomalyDto] {
        let url = URL(string: "\(baseUrl)anomalies")!
        let request = authorizedRequest(url: url)
        let (data, _) = try await URLSession.shared.data(for: request)
        return try JSONDecoder().decode([PriceAnomalyDto].self, from: data)
    }

    func fetchWeatherAlerts() async throws -> WeatherAlertDto {
        let url = URL(string: "\(baseUrl)weather-alerts")!
        let request = authorizedRequest(url: url)
        let (data, _) = try await URLSession.shared.data(for: request)
        return try JSONDecoder().decode(WeatherAlertDto.self, from: data)
    }

    func fetchBudgetRecipes() async throws -> [BudgetRecipeDto] {
        let url = URL(string: "\(baseUrl)budget-recipes")!
        let request = authorizedRequest(url: url)
        let (data, _) = try await URLSession.shared.data(for: request)
        return try JSONDecoder().decode([BudgetRecipeDto].self, from: data)
    }

    func fetchForecast(produceId: String) async throws -> PricePredictionDto {
        let url = URL(string: "\(baseUrl)forecast/\(produceId)")!
        let request = authorizedRequest(url: url)
        let (data, _) = try await URLSession.shared.data(for: request)
        return try JSONDecoder().decode(PricePredictionDto.self, from: data)
    }

    func fetchSeasonal() async throws -> [ProduceDto] {
        let url = URL(string: "\(baseUrl)seasonal")!
        let request = authorizedRequest(url: url)
        let (data, _) = try await URLSession.shared.data(for: request)
        return try JSONDecoder().decode([ProduceDto].self, from: data)
    }

    func fetchFavorites() async throws -> [FavoriteAlertDto] {
        let url = URL(string: "\(baseUrl)favorites")!
        let request = authorizedRequest(url: url)
        let (data, _) = try await URLSession.shared.data(for: request)
        return try JSONDecoder().decode([FavoriteAlertDto].self, from: data)
    }

    func deleteFavorite(produceId: String) async throws {
        let url = URL(string: "\(baseUrl)favorites/\(produceId)")!
        var request = authorizedRequest(url: url)
        request.httpMethod = "DELETE"
        let _ = try await URLSession.shared.data(for: request)
    }

    func submitCommunityPrice(cropCode: String, retailPrice: Double, market: String) async throws {
        let url = URL(string: "\(baseUrl)community-price")!
        var request = authorizedRequest(url: url)
        request.httpMethod = "POST"
        request.setValue("application/json", forHTTPHeaderField: "Content-Type")
        let body: [String: Any] = ["cropCode": cropCode, "retailPrice": retailPrice, "market": market]
        request.httpBody = try JSONSerialization.data(withJSONObject: body)
        let _ = try await URLSession.shared.data(for: request)
    }

    func fetchCommunityPrices(cropCode: String) async throws -> [CommunityPriceDto] {
        let url = URL(string: "\(baseUrl)community-price/\(cropCode)")!
        let request = authorizedRequest(url: url)
        let (data, _) = try await URLSession.shared.data(for: request)
        return try JSONDecoder().decode([CommunityPriceDto].self, from: data)
    }
}

File: native/Android/app/src/main/java/com/example/produceapp/data/ProduceService.kt
package com.example.produceapp.data

import retrofit2.http.*

/**
 * Retrofit 網路請求介面 - 定義所有後端 API 端點
 */
interface ProduceService {

    // === 認證 ===
    @POST("auth/token")
    suspend fun getAuthToken(@Body request: AuthTokenRequest): AuthTokenResponse

    // === 今日菜價 ===
    @GET("daily-prices")
    suspend fun getDailyPrices(
        @Query("keyword") keyword: String? = null,
        @Query("page") page: Int = 1,
        @Query("pageSize") pageSize: Int = 20
    ): PaginatedResult<ProduceDto>

    // === 歷史價格 ===
    @GET("history/{produceId}")
    suspend fun getPriceHistory(@Path("produceId") produceId: String): List<HistoricalPriceDto>

    // === 熱門交易 ===
    @GET("top-volume")
    suspend fun getTopVolume(): List<ProduceDto>

    // === 價格異常 ===
    @GET("anomalies")
    suspend fun getAnomalies(): List<PriceAnomalyDto>

    // === 價格預測 ===
    @GET("forecast/{produceId}")
    suspend fun getForecast(@Path("produceId") produceId: String): PricePredictionDto

    // === 當季盛產 ===
    @GET("seasonal")
    suspend fun getSeasonalCrops(): List<SeasonalCropDto>

    // === 天氣警報 ===
    @GET("weather-alerts")
    suspend fun getWeatherAlerts(): WeatherAlertDto

    // === 省錢食譜 ===
    @GET("budget-recipes")
    suspend fun getBudgetRecipes(): List<BudgetRecipeDto>

    // === 市場比價 ===
    @GET("compare/{cropName}")
    suspend fun comparePrices(@Path("cropName") cropName: String): List<ProduceDto>

    // === 收藏管理 ===
    @GET("favorites")
    suspend fun getFavorites(): List<FavoriteAlertDto>

    @DELETE("favorites/{produceId}")
    suspend fun removeFavorite(@Path("produceId") produceId: String)

    // === 社群回報 ===
    @POST("community-price")
    suspend fun submitCommunityPrice(@Body dto: CommunityPriceSubmitDto)

    @GET("community-price/{cropCode}")
    suspend fun getCommunityPrices(@Path("cropCode") cropCode: String): List<CommunityPriceDto>

    // === FCM Token ===
    @POST("fcm/token")
    suspend fun registerFcmToken(@Body request: FcmTokenRequest)
}
